### ISSUE TITLE
how can i get GPU usage

### ISSUE THREAD

**Issue by dwj1979** [oshi/oshi#1248]
how can i get GPU usage

**Comment by dbwiddis** [comment]
OSHI currently doesn't support this information. It requires newer hardware and compatible manufacturer-specific drivers, and isn't really suited to a cross-platform API at the current state of technology.

**Comment by dbwiddis** [comment]
To be more specific:

OSHI is a cross-platform Java library, intended to provide general cross-platform statistics.

GPUs are highly specialized hardware supported by different manufacturers, and there is not yet a consistent standard for accessing information on them.   

 - For Nvidia cards there is `nvidia-smi` tool (Linux only) which gives a whole bunch of formatting around a couple bits of information, and is not fully supported by many chips
 - For Intel there is `intel-gpu-tools` (Linux only) which provides more easily parsable information but it's still limited
 - For AMD there is `fglrx` (Linux only) which requires closed source drivers. Open source drivers have minimal information.

And that's just Linux.

macOS has some info in the IO Registry (IOAccellerator classes) but it's different for Intel and Radeon and has a whole bunch of 0's, usually only RAM % is consistently reliable and even that is guessing sometimes.

Windows requires a proprietary manufacturer-specific DLL to download, and lots of their graphics API is C++ only and not usable via JNA.

Bottom line:  if I put in a ton of work I could *maybe* get a "% memory" statistic working on a fraction of machines.  Compare this to very detailed software available for individuals to download for their specific OS and hardware that provide so much more useful information.

So, this just isn't going to be part of OSHI.  IF you want GPU usage for your specific GPU hardware and your specific Operating System, I'm sure there's a few options out there for you!

If you want a general tool to collect information about any GPU on any OS, such a thing likely doesn't exist.

**Comment by dwj1979** [comment]
thanks for answer. now i'a using c++ collect the GPU infomation on windows platform, and then send the information to my business system(devlope by java).  I found OSHI can gether lots of infomations, it is very good tools. so if OSHI can gether GPU uage, it's perfact!

**Comment by dbwiddis** [comment]
This was previously requested in #400 and I did spend some time looking at the options.

Not sure what Windows API you're using. Can you share?  It looks like NVIDIA and AMD both have APIs for their most recent cards, but not older ones. Intel explicitly says they don't support it.  There is a hidden Windows API that the sysinternals "Process Explorer" tool uses that some people have managed to hack together some sort of solution with, but I'm very reluctant to put undocumented APIs into this library.

MacOS has options reading the IORegistry and is the easiest of the options. However, it has multiple integrated graphics cards and the data only reports for one of them.

For linux, the only options appear to be parsing command lines of three separate programs by different GPU manufacturers.

All just to get a "% usage" number.

There are so many better options dedicated to specific hardware that users can and should leverage for this information.  I see very little reason to try to do this from inside Java in a cross-platform manner.

If you want to put forth all that work and submit a PR to the project, I'd be more than happy to consider it!  However, as a free project, this falls into the "juice is not worth the squeeze" category, at least until there is some semblance of cross-platform standards for GPU information.

**Comment by dwj1979** [comment]
Thank you for having taked the time to answer and explained me.

**Comment by Osiris-Team** [comment]
@dbwiddis Any updates since then?
Found [this](https://stackoverflow.com/questions/29611611/c-windows-gpu-usage-load):
![image](https://user-images.githubusercontent.com/59899645/177866123-09d05ec5-3d46-48d6-80af-4747ecdbbcb6.png)

**Comment by dbwiddis** [comment]
As stated in an earlier comment:
> Bottom line: if I put in a ton of work I could maybe get a "% memory" statistic working on a fraction of machines. Compare this to very detailed software available for individuals to download for their specific OS and hardware that provide so much more useful information.

You've pointed to WIndows code to get % GPU.  That's good.  If there's something similar for macOS and Linux that doesn't take up too much effort, it can be added.  I'll even help whoever wants to add it write the code. :)

**Comment by dbwiddis** [comment]
Looks like there's "Performance Statistics" in the [IO Registry](https://stackoverflow.com/questions/10110658/programmatically-get-gpu-percent-usage-in-os-x) that we can leverage for macOS.

**Comment by dbwiddis** [comment]
For Linux looks like you just trial-and-error [up to 4 command line tools](https://askubuntu.com/a/387600).  Which is the hardest part of doing this and the part that I'd need people with those GPU cards to test.

**Comment by dbwiddis** [comment]
So yeah, even after revisiting this.  I could get usage (and maybe memory) values from Windows or macOS. 

But compare the effort to get numbers in Linux vs. the massive # of tools already out there.  It's a ton of work and nearly impossible for me to test.

**Comment by Osiris-Team** [comment]
Yeah its sad that its gotta be so hard just for a percentage value.
Found another tool that does, and actually already it in Java: https://github.com/profesorfalken/jSensors

**Comment by dbwiddis** [comment]
Yep, it's using `libsensors`.  OSHI has long advertised that you don't need to install any additional DLLs, although certainly if someone has installed a package on Linux it could be used as an optional command-line target.  

The Windows stuff comes from OpenHardwareMonitor which OSHI does use for temperature/fan/voltage on Windows if it's running.

Again... if someone else wants to do the work I'll be happy to include all this and even guide the implementation.  WIndows and macOS seem reasonable, and Linux could be parsing command lines of a few of the most popular programs.   It is a lot of work for just a few bits of data (% Usage, some memory stats).  So it's not work I personally will do. :)

In contrast, the official GPU utilities are extremely capable and what any serious GPU user probably wants to use anyway... and they only need the one that matches their hardware.

**Comment by Osiris-Team** [comment]
I was going to go the easy way and use jSensors, but it looks like that it doesn't work...
